pkg/operator/controller/cabundle-configmap/cabundle_configmap.go

```diff
@@ -0,0 +1,299 @@
+// identification easier.
+func encodeSerialNumber(snum big.Int) string {
+	var newStrSNum string
+	strSNum := hex.EncodeToString(snum.Bytes())
+	for i := 0; i < len(strSNum); i++ {
+		if i != 0 && i%2 == 0 {
+			newStrSNum += ":"
+		}
+		newStrSNum += string(strSNum[i])
+	}
+	return newStrSNum
+}
+
+// getCertificatePrintId returns an identifier of the format <Certificate SerialNumber>-
+// <Certificate CommonName> which is used in logging and for better identification of
+// certificate in the bundle.
+func getCertificatePrintId(cert *x509.Certificate) string {
+	return fmt.Sprintf("%s(SerialNumber)-%s(CommonName)", encodeSerialNumber(*cert.SerialNumber), cert.Subject.CommonName)
+}
+
+// validateCACertificateBundle validates all the certificates present in the CA bundle.
+func validateCACertificateBundle(caCertBundleName string, caCertBundle []byte, certBuffer *bytes.Buffer) (int, int, error) {
+	certData := make([]certificateData, 0)
+	totalParsedCerts, totalValidCerts := 0, 0
+	for block, rest := pem.Decode(caCertBundle); block != nil; block, rest = pem.Decode(rest) {
+		switch block.Type {
+		case "CERTIFICATE":
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return totalParsedCerts, totalValidCerts,
+					fmt.Errorf("failed to parse CA certificate in %s bundle: %w", caCertBundleName, err)
+			}
+			totalParsedCerts++
+			certID := getCertificatePrintId(cert)
+			if skip := validateCACertificate(cert, caCertBundleName, certID, &certData); !skip {
+				if err := pem.Encode(certBuffer, block); err != nil {
+					return totalParsedCerts, totalValidCerts,
+						fmt.Errorf("failed to encode %s certificate of %s bundle: %w", certID, caCertBundleName, err)
+				}
+				totalValidCerts++
+			}
+		default:
+			log.Error(nil, "CA certificate bundle contains non-certificate data, discarded", "ca bundle name", caCertBundleName, "type", block.Type)
+			continue
+		}
+	}
+	return totalParsedCerts, totalValidCerts, nil
+}
+
+// validateCACertificate does below validations on a CA certificate.
+// - Secure Signature Algorithm is made use of for certificate signing.
+// - Certificate Validity.
+// - Certificate is a Certificate Authority.
+// - Looks for repetitive/duplicate CA certificates in the bundle.
+func validateCACertificate(cert *x509.Certificate, caCertBundleName, certID string, validatedCerts *[]certificateData) bool {
+	skip := false
+	if _, ok := insecureCertificateSignatureAlgorithms[cert.SignatureAlgorithm]; ok {
+		log.Error(nil, "certificate uses insecure signature algorithm, discarded", "certificate", certID, "signature algorithm", cert.SignatureAlgorithm, "certificate bundle", caCertBundleName)
+		skip = true
+	}
+
+	curTime := time.Now()
+	if curTime.Sub(cert.NotBefore).Hours() < 0 {
+		log.Info("certificate not yet valid, but will be considered", "certificate", certID, "valid after", cert.NotBefore.String(), "certificate bundle", caCertBundleName)
+	}
+	if curTime.Sub(cert.NotAfter).Hours() > 0 {
```

Comment:
Router returns the common error, but yes, if the HAProxy debug logs are enabled a user can find out the exact SSL error, and it would create confusion, but I think we could cover this in docs? or should we just log the error and continue to keep the CA certificate?

Reply:
I feel like it might be safer to err on the side of leaving the cert / logging the error and keeping the HaProxy failure mode of an expired CA Certificate. As long as the expired CA cert doesn't crash HaProxy (I'm pretty confident it doesn't), then I'm not totally sure what the net advantage is of skipping/removing it. HaProxy should handle expired CA's in a reasonable way.

Reply:
If HAProxy does not error out with the expired CA certificate, we should keep it.  The cluster-admin expressed intent to use the CA certificate, so we should use it if possible (that is, if doing using it doesn't break ingress by causing HAProxy to refuse to start).  However, it wouldn't be a bad idea to warn the cluster-admin about the expired CA certificate in an alert.

Reply:
suggestion incorporated. Just the certificates with insecure signature algorithm and non-ca certificates will be skipped now.